### PR TITLE
Battle Arena

### PR DESCRIPTION
**Pull request by GarOfMeridian**
- Jackpot capabilities - splits jackpot based on highest killer, winning
  and losing team. Players may drop shillings (over 10k) to 
  add to the jackpot. This should be the primary method of 
  encouraging competition - heavily discourage, possibly even mute,
  those who broadcast spam about "anyone want to battle arena".
- Notifications on / off - turn off global messages from this arena.
  This includes battle initiation (5 minutes to get to arena)
  and jackpot boosts.
- Placed in Tos (use new back door of Arena). Style is consistent
  with existing Arena architecture. This arena is underground,
  underneath the existing one.
- Appropriate minimap dots for combatants.
- Spectator system updated.

---

In this Battle Arena, players may say 'begin countdown' to initiate a fight. The server is told (all those who don't have their notifications from here turned off) and players have 5 minutes to show up. Kills and area control contribute to point totals. When the fight ends, the jackpot is split unevenly among highest killer, winners, and losers.

Player with the most kills gets 10% of the jackpot first.

Then everyone (including highest killer) splits 150% of an even share (winning team members) and 50% of an even share (losing team members). For example, a jackpot of 500k would immediately award 50k to the highest killer, then split 450k among the players. If there are ten players, the winners get 67.5k and the losers get 22.5k each (150% and 50% of 450k/10). This money almost always comes from players who contribute by dropping shilling amounts in the Battle Arena.

You cannot 'create' money this way. The Arena saves initial player counts. If you leave the battle early, you just waste your possible payout.

Sometimes events will boost the jackpot to encourage players to compete. That is the only time 'free money' becomes part of the reward. Ideally, we should have this scheduled to automatically happen every Friday or a Saturday 2-3 times. That way, the Arena is always open, and players can have battles any time, but 'battle arena events' still happen in a scheduled manner that people can plan to attend ahead of time.

---

Possible Todos:
- Do we need an ongoing point system to accompany this? How would we even record teams, since they are randomly created? Perhaps just highest kill counts overall?

### CODE REVIEW

File: kod/include/blakston.khd [hunk 1]
```diff
@@ -922,9 +922,11 @@
    RID_OLD_BAR_NORTH = 9002
    RID_OLD_BAR_SOUTH = 9003
   
-   % Survival Rooms 9500 - 9999
+   % Survival Rooms 9500 - 9899
    RID_SURVIVAL_START = 9500
-   RID_SURVIVAL_END = 9999
+   RID_SURVIVAL_END = 9899
+   
+   RID_BATTLE_ARENA = 9900
```
Comment (skittles1) [inline kod/include/blakston.khd]: Odd choice of room number - if it's in Tos, use the RIDs near Tos. 77 is free.

**Comment by skittles1** [comment]
Can't put line comments on battlearena.kod so going to have to put them here.
- All the explanation info should be on a sign instead, and add a new message given to every user who enters with something like "Welcome to the battle arena, read the sign for more information!". Then you can remove all the explanation timer stuff and the property from user.kod. If we want users to be able to toggle global messages, we can add that as a preference but as with public survival, I'd be inclined to leave this always-on so people are aware the arena exists. We can't force players to use it, but we can tempt them with messages of it being used.
- For the kill counts, you can use the util.kod scoreboard message which keeps the list sorted, however this message is changing names in the next merge to develop so perhaps hold off on reading it.
- If the purpose of the Battle Control radius spell is just to light up and check if a user is inside the circle, you can do the same thing with a dynamic light and the GetLocationInfoBSP C call (which gives you the ID of the sector). Using the C call here would also fix a rare bug, where the user enters or leaves the totem area and the 1 second control timer goes off before the 250ms RecalculateActiveEnchantments timer updates the effect.

Please also use the checking functions in the room editor, battlearena1.roo has a few issues.

**Comment by GarOfMeridian** [comment]
Ran those checks, does sector -1 need to be closed? If so, how?

**Comment by Maikeljr** [comment]
I like the idea, I just want to point out if anyone thought of what happens when you lose or leave the battle? I assume stands similar to other arena? Reasoning is pkers will simply chill outside looking for easy picks, can we buy rescue pots from an npc there or do we need to bring our own? Just a couple thoughts. Also buffing before or during combat? Is there a possible buff only section of battle? Its always awkward having people buff before vs mid combat.

**Comment by TheDiggie** [comment]
We can have the arena remove all buffs from players upon starting the arena. Fair fight for everyone

**Comment by M59Gar** [comment]
Updates have been made. I'm going to leave buffs / prebuffing up to players. This game mode is going to involve LOTS of dying. We'll see if prebuffing even accomplishes anything. As for PKs camping outside, the Tos Arena has always had that problem :) if it's super problematic, something can certainly be done.

**Pull request merged**